src/aap_eda/api/serializers/event_stream.py

```
@@ -0,0 +1,218 @@
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import logging
+import uuid
+
+import yaml
+from django.conf import settings
+from django.core.validators import RegexValidator
+from rest_framework import serializers
+from rest_framework.exceptions import ValidationError
+
+from aap_eda.api.constants import (
+    PG_NOTIFY_TEMPLATE_RULEBOOK_DATA,
+    PG_NOTIFY_TEMPLATE_RULEBOOK_NAME,
+)
+from aap_eda.api.exceptions import (
+    MissingEventStreamRulebook,
+    MissingEventStreamRulebookKeys,
+)
+from aap_eda.api.serializers.credential import CredentialSerializer
+from aap_eda.api.serializers.utils import substitute_extra_vars, swap_sources
+from aap_eda.core import models, validators
+
+logger = logging.getLogger(__name__)
+
+EDA_CHANNEL_PREFIX = "eda_"
+
+
+def _get_rulebook():
+    rulebook = None
+    name = settings.PG_NOTIFY_TEMPLATE_RULEBOOK
+    if name:
+        rulebook = models.Rulebook.objects.filter(name=name).first()
+
+        if not rulebook:
+            logger.error(
+                "Missing Listener rulebook %s",
+                settings.PG_NOTIFY_TEMPLATE_RULEBOOK,
+            )
+            raise MissingEventStreamRulebook
+
+        required_keys = ["type", "name", "args"]
+        rulesets = yaml.safe_load(rulebook.rulesets)
+        for ruleset in rulesets:
+            sources = ruleset.get("sources", [])
+            for source in sources:
+                complementary_source = source.get("complementary_source")
```

@hsong-rh if complementary_source is missing we should raise an Exception.